postgresql/resource_postgresql_security_label_test.go

```diff
@@ -0,0 +1,214 @@
+package postgresql
+
+import (
+	"database/sql"
+	"fmt"
+	"strings"
+	"testing"
+
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/resource"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/terraform"
+)
+
+func TestAccPostgresqlSecurityLabel_Basic(t *testing.T) {
+	resource.Test(t, resource.TestCase{
+		PreCheck: func() {
+			testAccPreCheck(t)
+			testCheckCompatibleVersion(t, featureSecurityLabel)
+			testSuperuserPreCheck(t)
+		},
+		Providers:    testAccProviders,
+		CheckDestroy: testAccCheckPostgresqlSecurityLabelDestroy,
+		Steps: []resource.TestStep{
+			{
+				Config: testAccPostgresqlSecurityLabelConfig,
+				Check: resource.ComposeTestCheckFunc(
+					testAccCheckPostgresqlSecurityLabelExists("postgresql_security_label.test_label"),
+					resource.TestCheckResourceAttr(
+						"postgresql_security_label.test_label", "object_type", "role"),
+					resource.TestCheckResourceAttr(
+						"postgresql_security_label.test_label", "object_name", "security_label_test_role"),
+					resource.TestCheckResourceAttr(
+						"postgresql_security_label.test_label", "label_provider", "dummy"),
+					resource.TestCheckResourceAttr(
+						"postgresql_security_label.test_label", "label", "secret"),
+				),
+			},
+		},
+	})
+}
+
+func TestAccPostgresqlSecurityLabel_Update(t *testing.T) {
+	resource.Test(t, resource.TestCase{
+		PreCheck: func() {
+			testAccPreCheck(t)
+			testCheckCompatibleVersion(t, featureSecurityLabel)
+			testSuperuserPreCheck(t)
+		},
+		Providers:    testAccProviders,
+		CheckDestroy: testAccCheckPostgresqlSecurityLabelDestroy,
+		Steps: []resource.TestStep{
+			{
+				Config: testAccPostgresqlSecurityLabelConfig,
+				Check: resource.ComposeTestCheckFunc(
+					testAccCheckPostgresqlSecurityLabelExists("postgresql_security_label.test_label"),
+					resource.TestCheckResourceAttr(
+						"postgresql_security_label.test_label", "object_type", "role"),
+					resource.TestCheckResourceAttr(
+						"postgresql_security_label.test_label", "object_name", "security_label_test_role"),
+					resource.TestCheckResourceAttr(
+						"postgresql_security_label.test_label", "label_provider", "dummy"),
+					resource.TestCheckResourceAttr(
+						"postgresql_security_label.test_label", "label", "secret"),
+				),
+			},
+			{
+				Config: testAccPostgresqlSecurityLabelChanges2,
+				Check: resource.ComposeTestCheckFunc(
+					testAccCheckPostgresqlSecurityLabelExists("postgresql_security_label.test_label"),
+					resource.TestCheckResourceAttr(
```

Comment:
Don't think it's necessary, a `Read` operation follows every create/update (aka, terraform apply operation), so in reality I don't think even `testAccCheckPostgresqlSecurityLabelExists` is needed, since the state will be updated by the Read operation after every apply.

Reply:
If you assume that the `Read` function works well indeed.

> I don't think even testAccCheckPostgresqlSecurityLabelExists is needed, since the state will be updated by the Read operation after every apply.

If you remove `testAccCheckPostgresqlSecurityLabelExists`, I can update the code and set the `resourcePostgreSQLSecurityLabelReadImpl` function to
```go
func resourcePostgreSQLSecurityLabelReadImpl(db *DBConnection, d *schema.ResourceData) error {
    d.SetId(generateSecurityLabelID(d))
    return nil
}
```
and the `resourcePostgreSQLSecurityLabelUpdateImpl` function to:
```go
func resourcePostgreSQLSecurityLabelUpdateImpl(db *DBConnection, d *schema.ResourceData, label string) error {
    return nil
}
```
and your tests are still passing (I tested)

Of course this example is very far-fetched but it's just to demonstrate than checking the state is not enough.

Reply:
Right, since we follow the rules to let `Read` function re-conciliate the TF state with the remote state, I think the current tests with `testAccCheckPostgresqlSecurityLabelExists` and `resource.TestCheckResourceAttr` covers what we intended to test.